<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork970-2023-01-01">
    </a>
</p>

# Data Visualization

Estimated time needed: **30** minutes

## Objectives

Después de terminar este lab seras capas de:

*   Visualizar Información con Python
*   Usar multiples librerías para visualizar información con Python


## Introducción
El objetivo de este laboratorio es introducirte a data visualization con Python de forma mas concreta y consistente posible. Hablando de consistencia, por que no hay una *mejor* librería de visualización disponible para python -hasta la creación de este lab- te introduciremos a diferentes librerías y te mostraremos sus beneficios cuando estaremos discutiendo nuevos conceptos de visualización. Haciendo esto, esperamos que el estudiante adquiera conocimientos variados sobre las bibliotecas y conceptos de visualización, de modo que puedan evaluar y decidir sobre la mejor técnica y herramienta de visualización para un problem y audiencia específicos

Por favor que has completado los pre requisitos para este curso,  [**Python Basics for Data Science**](https://www.edx.org/course/python-basics-for-data-science-2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01) and [**Analyzing Data with Python**](https://www.edx.org/course/data-analysis-with-python?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01).

**Nota**: La mayoría de los gráficos y las visualizaciones sera generadas usando información almacenada en dataframes de pandas. Por lo tanto, en este laboratorio, Proporcionaremos un pequeño curso introductorio de pandas. Si te interesa aprender mas sobre la librería pandas, la descripción detallada y explicación de como se usa y como se limpia, la jerarquía, y procesamiento de la información almacenada en pandas son entregadas en este curso [**Analyzing Data with Python**](https://www.edx.org/course/data-analysis-with-python?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01).

# Explorando Datasets con *pandas* <a id="0"></a>

*pandas* es una herramienta esencial para el análisis de datos en Python. Desde su [website](http://pandas.pydata.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01):

> *pandas* es un paquete de Python que provee una estructura de datos rápida, flexible y expresiva diseñada para trabajar con información "relational" o "labeled". Apunta a ser fundamental el bloque de alto nivel para data análisis practico en el **mundo real**.

El curso depende fuertemente de pandas *pandas* para la manipulación de datos, análisis, y visualización . Te alentamos fuertemente a que inviertas un poco de tu tiempo para familiarizarte con *pandas* API Reference: [http://pandas.pydata.org/pandas-docs/stable/api.html](http://pandas.pydata.org/pandas-docs/stable/api.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01).


## The Dataset: Immigration a Canada desde 1980 hasta 2013 <a id="2"></a>


Fuente del Dataset : [International migration flows to and from selected countries - The 2015 revision](http://www.un.org/en/development/desa/population/migration/data/empirical2/migrationflows.shtml?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01).

El dataset contiene información anual del flujo de inmigrantes internacionales registrados por los países de destino. La información presenta ambos flujos entradas a los países y salidas de a cuerdo al país de nacimiento, ciudadanía o lugar previo y futuro de residencia para ambos extranjeros y nacionales. La actual versión contiene la información de 45 países.

En este lab, nos enfocaremos en la información de inmigración de Canada.

![Data Preview](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/Module%201/images/DataSnapshot.png)

The Canada Immigration dataset can be fetched from <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01">here</a>.

***


## Bases de *pandas* <a id="4"></a>

Lo primero que debemos hacer es instalar **openpyxl** (antes llamado **xlrd**), un modulo que *pandas* requiere para leer archivos Excel.

In [75]:
#import piplite
#await piplite.install(['openpyxl==3.0.9'])

In [76]:
import numpy as np # Útil para diferentes areas de la ciencia de la computación en python
import pandas as pd # Librería primaria de la estructura de datos

Vamos a descargar e importar nuestro data set primario de inmigración Canadiense usando el método de **pandas** read_excel().

In [77]:
import urllib.request
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx'
resp = urllib.request.urlopen(URL)


if resp.getcode()==200:
    content = resp.read()
    df_can = pd.read_excel(io.BytesIO(content),
        sheet_name='Canada by Citizenship',
        skiprows=range(20),
        skipfooter=2)
    print('Data downloaded and read into a dataframe!')
else:
    print("Error al descargar el archivo")

Data downloaded and read into a dataframe!


Veamos las primeras 5 filas del dataset usando la función .head()

In [78]:
df_can.head()

Type    Coverage          OdName  AREA AreaName   REG  \
0  Immigrants  Foreigners     Afghanistan   935     Asia  5501   
1  Immigrants  Foreigners         Albania   908   Europe   925   
2  Immigrants  Foreigners         Algeria   903   Africa   912   
3  Immigrants  Foreigners  American Samoa   909  Oceania   957   
4  Immigrants  Foreigners         Andorra   908   Europe   925   

           RegName  DEV             DevName  1980  ...  2004  2005  2006  \
0    Southern Asia  902  Developing regions    16  ...  2978  3436  3009   
1  Southern Europe  901   Developed regions     1  ...  1450  1223   856   
2  Northern Africa  902  Developing regions    80  ...  3616  3626  4807   
3        Polynesia  902  Developing regions     0  ...     0     0     1   
4  Southern Europe  901   Developed regions     0  ...     0     0     1   

   2007  2008  2009  2010  2011  2012  2013  
0  2652  2111  1746  1758  2203  2635  2004  
1   702   560   716   561   539   620   603  
2  3623  4005  5393  4752  4325  3774  4331  
3     0     0     0     0     0     0     0  
4     1     0     0     0     0     1     1  

[5 rows x 43 columns]

También podemos ver las 5 filas del final con la función .tail()

In [79]:
df_can.tail()

Type    Coverage          OdName  AREA AreaName  REG  \
190  Immigrants  Foreigners        Viet Nam   935     Asia  920   
191  Immigrants  Foreigners  Western Sahara   903   Africa  912   
192  Immigrants  Foreigners           Yemen   935     Asia  922   
193  Immigrants  Foreigners          Zambia   903   Africa  910   
194  Immigrants  Foreigners        Zimbabwe   903   Africa  910   

                RegName  DEV             DevName  1980  ...  2004  2005  2006  \
190  South-Eastern Asia  902  Developing regions  1191  ...  1816  1852  3153   
191     Northern Africa  902  Developing regions     0  ...     0     0     1   
192        Western Asia  902  Developing regions     1  ...   124   161   140   
193      Eastern Africa  902  Developing regions    11  ...    56    91    77   
194      Eastern Africa  902  Developing regions    72  ...  1450   615   454   

     2007  2008  2009  2010  2011  2012  2013  
190  2574  1784  2171  1942  1723  1731  2112  
191     0     0     0     0     0     0     0  
192   122   133   128   211   160   174   217  
193    71    64    60   102    69    46    59  
194   663   611   508   494   434   437   407  

[5 rows x 43 columns]

Cuando analizamos un dataset, siempre es una buena idea comenzar por obtener la información básica de tu dataframe. Podemos hacer esto al usar el método .info().

Este método puede ser usado para obtener un sumario corto del dataframe.

In [80]:
df_can.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Columns: 43 entries, Type to 2013
dtypes: int64(37), object(6)
memory usage: 65.6+ KB


Para obtener la lista de las columnas headers podemos llamar a la variable de instancia columns en el dataframe

In [81]:
df_can.columns

Index([    'Type', 'Coverage',   'OdName',     'AREA', 'AreaName',      'REG',
        'RegName',      'DEV',  'DevName',       1980,       1981,       1982,
             1983,       1984,       1985,       1986,       1987,       1988,
             1989,       1990,       1991,       1992,       1993,       1994,
             1995,       1996,       1997,       1998,       1999,       2000,
             2001,       2002,       2003,       2004,       2005,       2006,
             2007,       2008,       2009,       2010,       2011,       2012,
             2013],
      dtype='object')

Similarmente para obtener la lista de los indices usamos la variable de instancia .index

In [82]:
df_can.index

RangeIndex(start=0, stop=195, step=1)

Nota: el tipo de variable por defecto en `index` y `columns` es **NOT** `list`.

In [83]:
print(type(df_can.columns))
print(type(df_can.index))

<class 'pandas.core.indexes.base.Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


Para obtener el indice de las columnas como lista, podemos usar el método .tolist().

In [84]:
df_can.columns.to_list()

['Type',
 'Coverage',
 'OdName',
 'AREA',
 'AreaName',
 'REG',
 'RegName',
 'DEV',
 'DevName',
 1980,
 1981,
 1982,
 1983,
 1984,
 1985,
 1986,
 1987,
 1988,
 1989,
 1990,
 1991,
 1992,
 1993,
 1994,
 1995,
 1996,
 1997,
 1998,
 1999,
 2000,
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013]

In [85]:
df_can.index.to_list()

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


In [86]:
print(type(df_can.columns.to_list()))
print(type(df_can.index.to_list()))

<class 'list'>
<class 'list'>


Para ver las dimensiones del dataframe, usamos la variable `shape`.

In [87]:
df_can.shape

(195, 43)

**Nota**: Los objetos principales almacenados en *pandas* son `float`, `int`, `bool`, `datetime64[ns]`, `datetime64[ns, tz]`, `timedelta[ns]`, `category` y `object` (string). ademas, los dtypes tienen items de tamaño, Ejemplo `int64` y `int32`.

In [88]:
# En pandas axis=0 representa las filas (por defecto) y axis=1 representa las columnas.

df_can.drop(["AREA","REG","DEV","Type","Coverage"], axis=1, inplace=True)
df_can.head(2)

OdName AreaName          RegName             DevName  1980  1981  \
0  Afghanistan     Asia    Southern Asia  Developing regions    16    39   
1      Albania   Europe  Southern Europe   Developed regions     1     0   

   1982  1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  2010  \
0    39    47    71   340  ...  2978  3436  3009  2652  2111  1746  1758   
1     0     0     0     0  ...  1450  1223   856   702   560   716   561   

   2011  2012  2013  
0  2203  2635  2004  
1   539   620   603  

[2 rows x 38 columns]

Vamos a renombrar las columnas para que tengan sentido. Podemos usar el método `rename()` le pasamos un diccionario de los antiguos nombres y los nuevos, de la siguiente manera:

In [89]:
df_can.rename(columns={"OdName":"Country","AreaName":"Continent","RegName":"Region"}, inplace=True)
df_can.columns

Index([  'Country', 'Continent',    'Region',   'DevName',        1980,
              1981,        1982,        1983,        1984,        1985,
              1986,        1987,        1988,        1989,        1990,
              1991,        1992,        1993,        1994,        1995,
              1996,        1997,        1998,        1999,        2000,
              2001,        2002,        2003,        2004,        2005,
              2006,        2007,        2008,        2009,        2010,
              2011,        2012,        2013],
      dtype='object')

Definimos las columnas para la suma, si no lo hacemos arrojara un error por que intentara sumar las primeras celdas que contienen strings


También agregaremos una columna "Total" que sume el total de inmigrantes por país por el periodo de 1980-2013, de la siguiente manera:

In [90]:
df_can["Total"]=df_can.iloc[:,4:].sum(axis=1)

Podemos revisar cuantos objetos null tenemos en el dataset, de la siguiente manera:

In [91]:
df_can.isnull().sum()

Country      0
Continent    0
Region       0
DevName      0
1980         0
1981         0
1982         0
1983         0
1984         0
1985         0
1986         0
1987         0
1988         0
1989         0
1990         0
1991         0
1992         0
1993         0
1994         0
1995         0
1996         0
1997         0
1998         0
1999         0
2000         0
2001         0
2002         0
2003         0
2004         0
2005         0
2006         0
2007         0
2008         0
2009         0
2010         0
2011         0
2012         0
2013         0
Total        0
dtype: int64

Finalmente, veamos un sumario rápido de cada columna en nuestro dataframe usando el método `describe()`.

In [92]:
df_can.describe()

1980          1981          1982          1983          1984  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     508.394872    566.989744    534.723077    387.435897    376.497436   
std     1949.588546   2152.643752   1866.997511   1204.333597   1198.246371   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%       13.000000     10.000000     11.000000     12.000000     13.000000   
75%      251.500000    295.500000    275.000000    173.000000    181.000000   
max    22045.000000  24796.000000  20620.000000  10015.000000  10170.000000   

              1985         1986          1987          1988          1989  \
count   195.000000   195.000000    195.000000    195.000000    195.000000   
mean    358.861538   441.271795    691.133333    714.389744    843.241026   
std    1079.309600  1225.576630   2109.205607   2443.606788   2555.048874   
min       0.000000     0.000000      0.000000      0.000000      0.000000   
25%       0.000000     0.500000      0.500000      1.000000      1.000000   
50%      17.000000    18.000000     26.000000     34.000000     44.000000   
75%     197.000000   254.000000    434.000000    409.000000    508.500000   
max    9564.000000  9470.000000  21337.000000  27359.000000  23795.000000   

       ...          2005          2006          2007          2008  \
count  ...    195.000000    195.000000    195.000000    195.000000   
mean   ...   1320.292308   1266.958974   1191.820513   1246.394872   
std    ...   4425.957828   3926.717747   3443.542409   3694.573544   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...     28.500000     25.000000     31.000000     31.000000   
50%    ...    210.000000    218.000000    198.000000    205.000000   
75%    ...    832.000000    842.000000    899.000000    934.500000   
max    ...  42584.000000  33848.000000  28742.000000  30037.000000   

               2009          2010          2011          2012          2013  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1275.733333   1420.287179   1262.533333   1313.958974   1320.702564   
std     3829.630424   4462.946328   4030.084313   4247.555161   4237.951988   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       36.000000     40.500000     37.500000     42.500000     45.000000   
50%      214.000000    211.000000    179.000000    233.000000    213.000000   
75%      888.000000    932.000000    772.000000    783.000000    796.000000   
max    29622.000000  38617.000000  36765.000000  34315.000000  34129.000000   

               Total  
count     195.000000  
mean    32867.451282  
std     91785.498686  
min         1.000000  
25%       952.000000  
50%      5018.000000  
75%     22239.500000  
max    691904.000000  

[8 rows x 35 columns]

***

## *pandas* Intermedio: Indexing y Selección (slicing)<a id="6"></a>

### Seleccionar una columna

**Hay dos formas de filtrar una columna por su nombre:**

Method 1: Rápido y fácil, pero solo funciona si el nombre de la columna NO tiene espacios o caracteres especialesQuick and easy, but only works if the column name does NOT have spaces or special characters.

```python
    df.column_name               # returns series
```

Method 2: Mas robusto y puede filtra a traves de multiples columnas

```python
    df['column']                  # devuelve una serie
```

```python
    df[['column 1', 'column 2']]  # devuelve un dataframe
```

***

Ejemplo:   Intentemos filtrar en la lista de countries ('Country').


In [93]:
df_can.Country

0         Afghanistan
1             Albania
2             Algeria
3      American Samoa
4             Andorra
            ...      
190          Viet Nam
191    Western Sahara
192             Yemen
193            Zambia
194          Zimbabwe
Name: Country, Length: 195, dtype: object

Vamos a filtrar la lista countries ('Country') y la información por años: 1980 - 1985.

In [94]:
df_can[["Country", 1980,1981,1982,1983,1984,1985]] # retorna un dataframe
# Observa que 'Country' es un string y los años son integers.
# Por el bien de la consistencia, convertiremos mas tarde todos los nombres de las columnas a string

Country  1980  1981  1982  1983  1984  1985
0       Afghanistan    16    39    39    47    71   340
1           Albania     1     0     0     0     0     0
2           Algeria    80    67    71    69    63    44
3    American Samoa     0     1     0     0     0     0
4           Andorra     0     0     0     0     0     0
..              ...   ...   ...   ...   ...   ...   ...
190        Viet Nam  1191  1829  2162  3404  7583  5907
191  Western Sahara     0     0     0     0     0     0
192           Yemen     1     2     1     6     0    18
193          Zambia    11    17    11     7    16     9
194        Zimbabwe    72   114   102    44    32    29

[195 rows x 7 columns]

### Seleccionar una Fila

Hay dos formas principales de seleccionar una fila:

```python
    df.loc[label]    # filtra por el rotulo del index o la columna
    df.iloc[index]   # filtrar por la posición del index o la columna
```


Antes de continuar, observa que el index por defecto de dataset es un numero que va desde el rango 0 al 194. Esto hace muy difícil hacer un query por un país en especifico. Por ejemplo para la información de Japón, necesitamos saber el valor del index correspondiente.

Esto puede ser fácilmente arreglado al usar la columna 'Country' como index usando el método `set_index()`.


In [95]:
df_can.set_index("Country",inplace=True) #Consejo: el opuesto de set es reset. Asi que para resetear el index podemos usar df_can.rest_index()

In [96]:
df_can.head(3)

Continent           Region             DevName  1980  1981  1982  \
Country                                                                        
Afghanistan      Asia    Southern Asia  Developing regions    16    39    39   
Albania        Europe  Southern Europe   Developed regions     1     0     0   
Algeria        Africa  Northern Africa  Developing regions    80    67    71   

             1983  1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  \
Country                              ...                                       
Afghanistan    47    71   340   496  ...  3436  3009  2652  2111  1746  1758   
Albania         0     0     0     1  ...  1223   856   702   560   716   561   
Algeria        69    63    44    69  ...  3626  4807  3623  4005  5393  4752   

             2011  2012  2013  Total  
Country                               
Afghanistan  2203  2635  2004  58639  
Albania       539   620   603  15699  
Algeria      4325  3774  4331  69439  

[3 rows x 38 columns]

In [97]:
# Opcional: Para remover el nombre del index.
df_can.index.name = None

Ejemplo: Veamos el numero de inmigrantes desde japón (fila 87) para los siguientes escenarios:

1\. La información completa de la fila (todas las columnas).

2\. Para el año 2013.

3\. Para los años 1980 al 1985


In [98]:
# 1. La información completa de la fila(todas las columnas)
df_can.loc["Japan"]

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [99]:
# Alternando métodos.
df_can.iloc[87]

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [100]:
df_can[df_can.index == "Japan"]

Continent        Region            DevName  1980  1981  1982  1983  \
Japan      Asia  Eastern Asia  Developed regions   701   756   598   309   

       1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  2011  2012  \
Japan   246   198   248  ...  1067  1212  1250  1284  1194  1168  1265  1214   

       2013  Total  
Japan   982  27707  

[1 rows x 38 columns]

In [101]:
# 2. Para el año 2013
df_can.loc["Japan",2013]

982

In [102]:
# Métodos alternativos
# El año 2013 es la ultima columna, con el index de posición 36
df_can.iloc[87,36]

982

In [103]:
# 3. para los años 1980 a 1985
df_can.loc["Japan",[1980,1981,1982,1983,1984]]

1980    701
1981    756
1982    598
1983    309
1984    246
Name: Japan, dtype: object

In [104]:
# Método alternativo.
df_can.iloc[87,[3,4,5,6,7,8]]

1980    701
1981    756
1982    598
1983    309
1984    246
1985    198
Name: Japan, dtype: object

Los nombres de las columnas que son integers (como los años) pueden inducir a una confusión. Por ejemplo, cuando nos referimos al año 2013, uno podría relacionar con la posición 2013 del index.

Para evitar esta ambigüedad, vamos a convertir los nombres de las columnas a string:'1980' al '2013'.


In [105]:
df_can.columns = list(map(str,df_can.columns)) 
[print(type(x))for x in df_can.columns.values]

<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>


[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None]

Dado que convertimos los años a string, vamos a declarar una variable que nos permitirá llamar completamente a la función range of years:


In [106]:
years = list(map(str,range(1980,2014)))
years

['1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013']

### Filtrando en base al criterio.

Para filtrar el dataframe basado en el criterio, simplemente pasamos la condición como un vector booleano.

Por ejemplo, vamos a filtrar el dataframe para mostrar la información de los países asiáticos (AreaName = Asia).

In [107]:
condition = df_can["Continent"] == "Asia"
print(condition)

Afghanistan        True
Albania           False
Algeria           False
American Samoa    False
Andorra           False
                  ...  
Viet Nam           True
Western Sahara    False
Yemen              True
Zambia            False
Zimbabwe          False
Name: Continent, Length: 195, dtype: bool


In [108]:
# 2 pasa esta condición al dataframe
df_can[condition]

Continent              Region  \
Afghanistan                                         Asia       Southern Asia   
Armenia                                             Asia        Western Asia   
Azerbaijan                                          Asia        Western Asia   
Bahrain                                             Asia        Western Asia   
Bangladesh                                          Asia       Southern Asia   
Bhutan                                              Asia       Southern Asia   
Brunei Darussalam                                   Asia  South-Eastern Asia   
Cambodia                                            Asia  South-Eastern Asia   
China                                               Asia        Eastern Asia   
China, Hong Kong Special Administrative Region      Asia        Eastern Asia   
China, Macao Special Administrative Region          Asia        Eastern Asia   
Cyprus                                              Asia        Western Asia   
Democratic People's Republic of Korea               Asia        Eastern Asia   
Georgia                                             Asia        Western Asia   
India                                               Asia       Southern Asia   
Indonesia                                           Asia  South-Eastern Asia   
Iran (Islamic Republic of)                          Asia       Southern Asia   
Iraq                                                Asia        Western Asia   
Israel                                              Asia        Western Asia   
Japan                                               Asia        Eastern Asia   
Jordan                                              Asia        Western Asia   
Kazakhstan                                          Asia        Central Asia   
Kuwait                                              Asia        Western Asia   
Kyrgyzstan                                          Asia        Central Asia   
Lao People's Democratic Republic                    Asia  South-Eastern Asia   
Lebanon                                             Asia        Western Asia   
Malaysia                                            Asia  South-Eastern Asia   
Maldives                                            Asia       Southern Asia   
Mongolia                                            Asia        Eastern Asia   
Myanmar                                             Asia  South-Eastern Asia   
Nepal                                               Asia       Southern Asia   
Oman                                                Asia        Western Asia   
Pakistan                                            Asia       Southern Asia   
Philippines                                         Asia  South-Eastern Asia   
Qatar                                               Asia        Western Asia   
Republic of Korea                                   Asia        Eastern Asia   
Saudi Arabia                                        Asia        Western Asia   
Singapore                                           Asia  South-Eastern Asia   
Sri Lanka                                           Asia       Southern Asia   
State of Palestine                                  Asia        Western Asia   
Syrian Arab Republic                                Asia        Western Asia   
Tajikistan                                          Asia        Central Asia   
Thailand                                            Asia  South-Eastern Asia   
Turkey                                              Asia        Western Asia   
Turkmenistan                                        Asia        Central Asia   
United Arab Emirates                                Asia        Western Asia   
Uzbekistan                                          Asia        Central Asia   
Viet Nam                                            Asia  South-Eastern Asia   
Yemen                                               Asia        Western Asia   

                                              

In [109]:
# Podemos pasar multiples criterios en una sola linea
# Vamos a filtrar por AreaNAme = Asia y RegName = Southern Asia

df_can[(df_can['Continent']=='Asia') & (df_can['Region']=='Southern Asia')]

# Nota:cuando usamos los operadores 'and' and 'or', pandas requiere que usemos '&' and '|' en vez de 'and' y 'or'
# No olvides encerrar tus condiciones en paréntesis.

Continent         Region             DevName  1980  \
Afghanistan                     Asia  Southern Asia  Developing regions    16   
Bangladesh                      Asia  Southern Asia  Developing regions    83   
Bhutan                          Asia  Southern Asia  Developing regions     0   
India                           Asia  Southern Asia  Developing regions  8880   
Iran (Islamic Republic of)      Asia  Southern Asia  Developing regions  1172   
Maldives                        Asia  Southern Asia  Developing regions     0   
Nepal                           Asia  Southern Asia  Developing regions     1   
Pakistan                        Asia  Southern Asia  Developing regions   978   
Sri Lanka                       Asia  Southern Asia  Developing regions   185   

                            1981  1982  1983  1984  1985  1986  ...   2005  \
Afghanistan                   39    39    47    71   340   496  ...   3436   
Bangladesh                    84    86    81    98    92   486  ...   4171   
Bhutan                         0     0     0     1     0     0  ...      5   
India                       8670  8147  7338  5704  4211  7150  ...  36210   
Iran (Islamic Republic of)  1429  1822  1592  1977  1648  1794  ...   5837   
Maldives                       0     0     1     0     0     0  ...      0   
Nepal                          1     6     1     2     4    13  ...    607   
Pakistan                     972  1201   900   668   514   691  ...  14314   
Sri Lanka                    371   290   197  1086   845  1838  ...   4930   

                             2006   2007   2008   2009   2010   2011   2012  \
Afghanistan                  3009   2652   2111   1746   1758   2203   2635   
Bangladesh                   4014   2897   2939   2104   4721   2694   2640   
Bhutan                         10      7     36    865   1464   1879   1075   
India                       33848  28742  28261  29456  34235  27509  30933   
Iran (Islamic Republic of)   7480   6974   6475   6580   7477   7479   7534   
Maldives                        0      2      1      7      4      3      1   
Nepal                         540    511    581    561   1392   1129   1185   
Pakistan                    13127  10124   8994   7217   6811   7468  11227   
Sri Lanka                    4714   4123   4756   4547   4422   3309   3338   

                             2013   Total  
Afghanistan                  2004   58639  
Bangladesh                   3789   65568  
Bhutan                        487    5876  
India                       33087  691904  
Iran (Islamic Republic of)  11291  175923  
Maldives                        1      30  
Nepal                        1308   10222  
Pakistan                    12603  241600  
Sri Lanka                    2394  148358  

[9 rows x 38 columns]

Antes de continuar, revisemos los cambios que hemos echo al dataframe.

In [110]:
print('Dimensiones de la información:',df_can.shape)
print(df_can.columns)
df_can.head(2)

Dimensiones de la información: (195, 38)
Index(['Continent', 'Region', 'DevName', '1980', '1981', '1982', '1983',
       '1984', '1985', '1986', '1987', '1988', '1989', '1990', '1991', '1992',
       '1993', '1994', '1995', '1996', '1997', '1998', '1999', '2000', '2001',
       '2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009', '2010',
       '2011', '2012', '2013', 'Total'],
      dtype='object')


Continent           Region             DevName  1980  1981  1982  \
Afghanistan      Asia    Southern Asia  Developing regions    16    39    39   
Albania        Europe  Southern Europe   Developed regions     1     0     0   

             1983  1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  \
Afghanistan    47    71   340   496  ...  3436  3009  2652  2111  1746  1758   
Albania         0     0     0     1  ...  1223   856   702   560   716   561   

             2011  2012  2013  Total  
Afghanistan  2203  2635  2004  58639  
Albania       539   620   603  15699  

[2 rows x 38 columns]

***

# Visualización de la información usando Matplotlib<a id="8"></a>

## Matplotlib: Librería estándar de visualización para Python<a id="10"></a>

La librería principal de representación gráfica que exploraremos en este curso es [Matplotlib](http://matplotlib.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01). Como menciona en su sitio:

> Matplotlib es una librería de gráficos 2D de Python lo que produce  Matplotlib puede utilizarse en scripts de Python, en el intérprete de Python y IPython, en el cuaderno Jupyter, en servidores de aplicaciones web y en cuatro entornos de interfaz gráfica de usuario (GUI). Esto permite producir gráficos de calidad de publicación en una variedad de formatos de impresión y en entornos interactivos en diferentes plataformas.

Si aspiras a crear una visualización que impacte con python, Matplotlib  es una herramienta esencial para tener a disposición.

### Matplotlib.Pyplot

Uno de los aspectos centrales de Matplotlib es `matplotlib.pyplot`. Es la capa de código de Matplotlib la cual estudiamos en los videos a cerca de Matplotlib. Recuerda que es una colección de comandos y funciones de estilo que hacen que Matplotlib trabaje MATLAB. Cada función de `pyplot` hace cambios a una figura: ejemplo., creas una figura, creas un area de representación gráfica en una figura, genera algunas lineas en el area de representación gráfica, decora el gráfico con rótulos, etc. en este lab, trabajaremos con la capa de código para aprender a generar gráficos de lineas. En los laboratorios futuros, trabajaremos con la capa artística y experimentaremos de primera mano cuan diferente es de la capa de código.

In [111]:
import matplotlib as mpl
import matplotlib.pyplot as plt

Opcional: revisa si Matplotlib esta cargado.

In [112]:
print("Versión de Matplotlib: ",mpl.__version__)

Versión de Matplotlib:  3.7.2


Opcional: Aplicar estilos a Matplotlib.

In [113]:
print(plt.style.available)
mpl.style.use(["ggplot"])

['Solarize_Light2', '_classic_test_patch', '_mpl-gallery', '_mpl-gallery-nogrid', 'bmh', 'classic', 'dark_background', 'fast', 'fivethirtyeight', 'ggplot', 'grayscale', 'seaborn-v0_8', 'seaborn-v0_8-bright', 'seaborn-v0_8-colorblind', 'seaborn-v0_8-dark', 'seaborn-v0_8-dark-palette', 'seaborn-v0_8-darkgrid', 'seaborn-v0_8-deep', 'seaborn-v0_8-muted', 'seaborn-v0_8-notebook', 'seaborn-v0_8-paper', 'seaborn-v0_8-pastel', 'seaborn-v0_8-poster', 'seaborn-v0_8-talk', 'seaborn-v0_8-ticks', 'seaborn-v0_8-white', 'seaborn-v0_8-whitegrid', 'tableau-colorblind10']


### Gráficar en *pandas*

Afortunadamente, pandas tiene una implementación built-in de Matplotlib que podemos usar. Gráfica in *pandas* es tan simple como añadir el método `.plot()` a una serie o dataframe.

Documentación:

*   [Plotting with Series](http://pandas.pydata.org/pandas-docs/stable/api.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01#plotting)<br>
*   [Plotting with Dataframes](http://pandas.pydata.org/pandas-docs/stable/api.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01#api-dataframe-plotting)


# Gráficos de lineas (Series/Dataframe) <a id="12"></a>

**Que es un gráfico de lineas y por que usarlo?**

Un gráfico de lineas el cual muestra información el cual muestra una serie de puntos llamados 'markers' conectados por una linea recta. Es un tipo básico de gráfico común en muchos campos.

Utiliza un gráfico de lineas cuando tengas un conjunto de datos continuos. Estos son más adecuados para visualizaciones basadas en tendencias de  datos a lo largo de un período de tiempo.

**Vamos a comenzar con un caso de estudio:**

En 2010, Haiti sufrió un catastrófico terremoto de magnitude 7.0. El terremoto causo una amplia devastación y perdida de vidas y cerca de tres millones de personas fueron afectadas por este desastre natural. Como parte del esfuerzo humanitario de Canada, el Gobierno de Canada aumento su esfuerzo en aceptar refugiados desde Haiti. Podemos visualizar rápidamente usando un gráfico `Line`:

**Pregunta:** Dibuja un gráfico de lineas de Haiti usando `df.plot()`.